portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -29,22 +40,53 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 //    response.setContentType("text/html;");
 //    response.getWriter().println("<h1>Hello world!</h1>");
 //    response.getWriter().println("<h1>Hello [Ruiyuan Qian]</h1>");
+//        String json = "{";
+//            json += "\"status\": ";
+//            json += 12345;
+//            json += "}";
+//        response.setContentType("application/json;");
+//        response.getWriter().println(json);
 
-String json = "{";
-    json += "\"status\": ";
-    json += 12345;
-    json += "}";
+    Query query = new Query("Memes").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    response.setContentType("text/html;");
 
-    response.setContentType("application/json;");
-    response.getWriter().println(json);
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String argURL = (String) entity.getProperty("url");
+      String argComment = (String) entity.getProperty("comment");
+      long timestamp = (long) entity.getProperty("timestamp");
+      
+      response.getWriter().println("\n <p> ### ### ### </p>\n");
+      response.getWriter().println("\n <p> ID: " + id + "</p>");
+      response.getWriter().println("\n <p> Timestamp: " + DateFormat.getDateTimeInstance().format(timestamp) + "</p>");
+      response.getWriter().println("\n <p> Comment: " + argComment + "</p>");
+      response.getWriter().println("\n <img src=\"" + argURL + "\"></img>");
```

Comment:
In general you should returns a list of json objects of all uploaded memes, then add a list_meme.html fetches via /data and parse Json in a XSS safe way.